File: src/errata/models/gemini.py
"""Google Gemini model adapter."""

from __future__ import annotations

import time
from collections.abc import Callable

from errata.models.base import AgentEvent, FileWrite, ModelAdapter, ModelResponse
from errata.tools import READ_TOOL_NAME, TOOL_DEFINITIONS, WRITE_TOOL_NAME, execute_read


def _build_gemini_tools():
    """Translate canonical tool defs to google.generativeai protos."""
    import google.generativeai as genai

    declarations = []
    for t in TOOL_DEFINITIONS:
        props = {}
        for name, spec in t["parameters"]["properties"].items():
            props[name] = genai.protos.Schema(
                type=genai.protos.Type.STRING,
                description=spec.get("description", ""),
            )
        declarations.append(
            genai.protos.FunctionDeclaration(
                name=t["name"],
                description=t["description"],
                parameters=genai.protos.Schema(
                    type=genai.protos.Type.OBJECT,
                    properties=props,
                    required=t["parameters"].get("required", []),
                ),
            )
        )
    return [genai.protos.Tool(function_declarations=declarations)]


class GeminiAdapter(ModelAdapter):
    def __init__(self, model_id: str, api_key: str) -> None:
        self.model_id = model_id
        self._api_key = api_key

    async def run_agent(
        self,
        prompt: str,
        on_event: Callable[[AgentEvent], None],
        verbose: bool = False,
    ) -> ModelResponse:
        import google.generativeai as genai

        genai.configure(api_key=self._api_key)
        model = genai.GenerativeModel(self.model_id, tools=_build_gemini_tools())
        chat = model.start_chat()

        proposed_writes: list[FileWrite] = []
        text_parts: list[str] = []
        start = time.monotonic()

        user_message = prompt

        while True:
            response = await chat.send_message_async(user_message)
            candidate = response.candidates[0]

            function_calls = []
            for part in candidate.content.parts:
                if part.text:
                    text_parts.append(part.text)
                    if verbose:
                        on_event(AgentEvent("text", part.text))
                if part.function_call.name:
                    function_calls.append(part.function_call)

            if not function_calls:
                break

            # Build function responses to feed back
            function_responses = []
            for fc in function_calls:
                args = dict(fc.args)
                if fc.name == READ_TOOL_NAME:
                    path = args["path"]
                    on_event(AgentEvent("reading", path))
                    content = execute_read(path)
                    function_responses.append(
                        genai.protos.Part(
                            function_response=genai.protos.FunctionResponse(
                                name=fc.name,
                                response={"result": content},
                            )
                        )
                    )
                elif fc.name == WRITE_TOOL_NAME:
                    path = args["path"]
                    on_event(AgentEvent("writing", path))
                    proposed_writes.append(FileWrite(path, args["content"]))
                    function_responses.append(
                        genai.protos.Part(
                            function_response=genai.protos.FunctionResponse(
                                name=fc.name,
                                response={"result": "Write queued — will be applied if selected."},
                            )
                        )
                    )

            user_message = genai.protos.Content(parts=function_responses, role="user")

        return ModelResponse(
            model_id=self.model_id,
            text="".join(text_parts),
            latency_ms=int((time.monotonic() - start) * 1000),
            proposed_writes=proposed_writes,
        )

File: internal/models/gemini.go
package models

import (
	"context"
	"time"

	"github.com/google/generative-ai-go/genai"
	"github.com/suarezc/errata/internal/tools"
	"google.golang.org/api/option"
)

// GeminiAdapter implements ModelAdapter for Gemini models.
type GeminiAdapter struct {
	modelID string
	apiKey  string
}

// NewGeminiAdapter creates a GeminiAdapter.
func NewGeminiAdapter(modelID, apiKey string) *GeminiAdapter {
	return &GeminiAdapter{modelID: modelID, apiKey: apiKey}
}

func (a *GeminiAdapter) ID() string { return a.modelID }

func (a *GeminiAdapter) RunAgent(
	ctx context.Context,
	prompt string,
	onEvent func(AgentEvent),
	verbose bool,
) (ModelResponse, error) {
	client, err := genai.NewClient(ctx, option.WithAPIKey(a.apiKey))
	if err != nil {
		return ModelResponse{ModelID: a.modelID, Error: err.Error()}, err
	}
	defer client.Close()

	model := client.GenerativeModel(a.modelID)
	model.Tools = buildGeminiTools()

	chat := model.StartChat()

	var textParts []string
	var proposed []tools.FileWrite
	var totalInput, totalOutput int64
	start := time.Now()

	userMsg := genai.Text(prompt)
	var currentMsg []genai.Part
	currentMsg = append(currentMsg, userMsg)

	for {
		resp, err := chat.SendMessage(ctx, currentMsg...)
		if err != nil {
			return ModelResponse{
				ModelID:      a.modelID,
				LatencyMS:    time.Since(start).Milliseconds(),
				InputTokens:  totalInput,
				OutputTokens: totalOutput,
				CostUSD:      CostUSD(a.modelID, totalInput, totalOutput),
				Error:        err.Error(),
			}, err
		}

		if resp.UsageMetadata != nil {
			totalInput += int64(resp.UsageMetadata.PromptTokenCount)
			totalOutput += int64(resp.UsageMetadata.CandidatesTokenCount)
		}

		candidate := resp.Candidates[0]
		var funcResponses []genai.Part
		currentMsg = nil

		for _, part := range candidate.Content.Parts {
			switch v := part.(type) {
			case genai.Text:
				textParts = append(textParts, string(v))
				if verbose {
					onEvent(AgentEvent{Type: "text", Data: string(v)})
				}

			case genai.FunctionCall:
				args := extractStringArgs(v.Args)
				switch v.Name {
				case tools.ReadToolName:
					path := args["path"]
					onEvent(AgentEvent{Type: "reading", Data: path})
					content := tools.ExecuteRead(path)
					funcResponses = append(funcResponses, genai.FunctionResponse{
						Name:     v.Name,
						Response: map[string]any{"result": content},
					})

				case tools.WriteToolName:
					path := args["path"]
					onEvent(AgentEvent{Type: "writing", Data: path})
					proposed = append(proposed, tools.FileWrite{Path: path, Content: args["content"]})
					funcResponses = append(funcResponses, genai.FunctionResponse{
						Name:     v.Name,
						Response: map[string]any{"result": "Write queued — will be applied if selected."},
					})
				}
			}
		}

		if len(funcResponses) == 0 {
			break
		}
		currentMsg = funcResponses
	}

	return ModelResponse{
		ModelID:        a.modelID,
		Text:           join(textParts),
		LatencyMS:      time.Since(start).Milliseconds(),
		InputTokens:    totalInput,
		OutputTokens:   totalOutput,
		CostUSD:        CostUSD(a.modelID, totalInput, totalOutput),
		ProposedWrites: proposed,
	}, nil
}

func buildGeminiTools() []*genai.Tool {
	var decls []*genai.FunctionDeclaration
	for _, def := range tools.Definitions {
		props := map[string]*genai.Schema{}
		for name, p := range def.Properties {
			props[name] = &genai.Schema{
				Type:        genai.TypeString,
				Description: p.Description,
			}
		}
		required := make([]string, len(def.Required))
		copy(required, def.Required)
		decls = append(decls, &genai.FunctionDeclaration{
			Name:        def.Name,
			Description: def.Description,
			Parameters: &genai.Schema{
				Type:       genai.TypeObject,
				Properties: props,
				Required:   required,
			},
		})
	}
	return []*genai.Tool{{FunctionDeclarations: decls}}
}

func extractStringArgs(args map[string]any) map[string]string {
	out := make(map[string]string, len(args))
	for k, v := range args {
		if s, ok := v.(string); ok {
			out[k] = s
		}
	}
	return out
}
